source/extensions/filters/network/thrift_proxy/router/router_impl.cc

```diff
@@ -24,24 +26,60 @@ RouteConstSharedPtr RouteEntryImplBase::clusterEntry() const { return shared_fro
 
 MethodNameRouteEntryImpl::MethodNameRouteEntryImpl(
     const envoy::config::filter::network::thrift_proxy::v2alpha1::Route& route)
-    : RouteEntryImplBase(route), method_name_(route.match().method()) {}
+  : RouteEntryImplBase(route),
+    method_name_(route.match().method_name()),
+    invert_(route.match().invert()) {
+  if (method_name_.empty() && invert_) {
+    throw EnvoyException("Cannot have an empty method name with inversion enabled");
+  }
+}
 
 RouteConstSharedPtr MethodNameRouteEntryImpl::matches(const MessageMetadata& metadata) const {
-  if (method_name_.empty()) {
+  bool matches = method_name_.empty() || (metadata.hasMethodName() && metadata.methodName() == method_name_);
+
+  if (matches ^ invert_) {
     return clusterEntry();
+  } else {
+    return nullptr;
   }
+}
 
-  if (metadata.hasMethodName() && metadata.methodName() == method_name_) {
-    return clusterEntry();
+ServiceNameRouteEntryImpl::ServiceNameRouteEntryImpl(
+    const envoy::config::filter::network::thrift_proxy::v2alpha1::Route& route)
+  : RouteEntryImplBase(route),
+    service_name_(route.match().service_name()),
+    invert_(route.match().invert()) {
+  if (service_name_.empty() && invert_) {
+    throw EnvoyException("Cannot have an empty service name with inversion enabled");
   }
+}
 
-  return nullptr;
+RouteConstSharedPtr ServiceNameRouteEntryImpl::matches(const MessageMetadata& metadata) const {
+  bool matches = service_name_.empty() ||
+    (metadata.hasMethodName() && StringUtil::startsWith(metadata.methodName().c_str(), service_name_));
```

Comment:
Need to make sure that a colon immediately follows service_name_. One way might be to initialize service_name_ with the colon appended, but only if it's not empty.